### PR TITLE
Fix/milstone-1, adds fixes for milstone-1 review 

### PR DESCRIPTION
**Pull request by davidkin**

### CODE REVIEW

File: config/config.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (danylkaaa) [inline config/config.json:1]: Треба подумати потім як діставати конфігурацію потім з .env

File: migrations/20230405145705-create-user.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                type: Sequelize.INTEGER
+            },
+            firstName: {
+                type: Sequelize.STRING
+            },
+            lastName: {
+                type: Sequelize.STRING
+            },
+            email: {
+                type: Sequelize.STRING,
```
Comment (danylkaaa) [inline migrations/20230405145705-create-user.js:19]: add index for this field

File: migrations/20230405145705-create-user.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                allowNull: false
+            },
+            role: {
+                type: Sequelize.ARRAY(Sequelize.STRING),
```
Comment (danylkaaa) [inline migrations/20230405145705-create-user.js:24]: role should be a separated table. Users with same emails may be drivers and riders

File: migrations/20230412105758-create-sessions.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            sid: {
+                type: Sequelize.STRING
+            },
+            userId: {
+                type: Sequelize.STRING
+            },
+            expires: {
+                type: Sequelize.DATE
+            },
+            data: {
```
Comment (danylkaaa) [inline migrations/20230412105758-create-sessions.js]: you don't need to store data in DB. The only purpose to have sessions table - to be able to manage signed tokens (show them to user and to be able to revoke)

File: migrations/20230412105758-create-sessions.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+                allowNull: false,
+                autoIncrement: true,
+                primaryKey: true,
+                type: Sequelize.INTEGER
+            },
+            sid: {
+                type: Sequelize.STRING
+            },
+            userId: {
+                type: Sequelize.STRING
```
Comment (danylkaaa) [inline migrations/20230412105758-create-sessions.js]: you need to specify it as foreign key and make index by this field

File: migrations/20230412105758-create-sessions.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            },
+            expires: {
+                type: Sequelize.DATE
```
Comment (danylkaaa) [inline migrations/20230412105758-create-sessions.js:20]: DATETIME?

File: migrations/20230412105758-create-sessions.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+            data: {
+                type: Sequelize.STRING
+            },
+            createdAt: {
+                allowNull: false,
+                type: Sequelize.DATE
+            },
+            updatedAt: {
+                allowNull: false,
+                type: Sequelize.DATE
```
Comment (danylkaaa) [inline migrations/20230412105758-create-sessions.js:28]: same

File: migrations/20230412105758-create-sessions.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+            },
+            expires: {
+                type: Sequelize.DATE
+            },
+            data: {
+                type: Sequelize.STRING
+            },
+            createdAt: {
+                allowNull: false,
+                type: Sequelize.DATE
```
Comment (danylkaaa) [inline migrations/20230412105758-create-sessions.js:24]: same

File: models/index.js
Comment (danylkaaa) [inline models/index.js:1]: this file is a dirty hack)
Read models manualy and register them in Sequalize

File: seeders/20230408144804-demo-user.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+/** @type {import('sequelize-cli').Migration} */
+module.exports = {
+    up: (queryInterface, Sequelize) => {
+        return queryInterface.bulkInsert('Users', [{
```
Comment (danylkaaa) [inline seeders/20230408144804-demo-user.js:6]: better create admin user, who can access system right after it was started

File: seeders/20230408144804-demo-user.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+            lastName: 'Doe',
+            email: 'example@example.com',
+            role: ['admin'],
+            password: '12345',
+            createdAt: new Date(),
+            updatedAt: new Date()
+        }]);
+    },
+    down: (queryInterface, Sequelize) => {
+        return queryInterface.bulkDelete('Users', null, {});
```
Comment (danylkaaa) [inline seeders/20230408144804-demo-user.js]: no, delete only 1 user

File: src/app.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }));
+
+app.use((req, res, next) => { requestContext.initStore(new ContextService(config, logger), next); });
+
+app.use(cors);
+app.use(httpLog);
+
+app.use('/', appRoutes);
```
Comment (danylkaaa) [inline src/app.ts]:
```suggestion
app.use(appRoutes);
```

File: src/config/config.ts [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+import { type ValidationResult } from 'joi';
+import { type AppConfig } from './config.interface';
+
+dotenv.config();
+
+const readConfig = (): AppConfig => {
+  const config: AppConfig = {
+    name: 'demo-uber',
+    port: 9000,
+    version: '1.0.0',
```
Comment (danylkaaa) [inline src/config/config.ts]: read from package.json

File: src/config/config.ts [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+import configSchema from './validation/config';
+import { type ValidationResult } from 'joi';
+import { type AppConfig } from './config.interface';
+
+dotenv.config();
+
+const readConfig = (): AppConfig => {
+  const config: AppConfig = {
+    name: 'demo-uber',
+    port: 9000,
```
Comment (danylkaaa) [inline src/config/config.ts]: read from .env

File: src/errors/AuthorizedError.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
+class NotAuthorizedError extends Error {
```
Comment (danylkaaa) [inline src/errors/AuthorizedError.ts]: check it out https://www.npmjs.com/package/@hapi/boom

File: src/middleware/http-middleware.ts [hunk 15]
```diff
@@ -1,8 +1,9 @@
 import { type NextFunction, type Request, type Response } from 'express';
+import logger from '../logger';
 
 const httpLog = (req: Request, res: Response, next: NextFunction): void => {
   res.on('finish', function () {
-    console.log(req.method, decodeURI(req.url), res.statusCode, res.statusMessage);
+    logger.info(req.method, decodeURI(req.url), res.statusCode, res.statusMessage);
```
Comment (danylkaaa) [inline src/middleware/http-middleware.ts:7]: get logger from context store

File: src/modules/routes.ts [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import express from 'express';
+import { type NextFunction, type Request, type Response } from 'express';
+import logger from '../logger';
+import userRoutes from './user/user.routes';
+import sessionRoutes from './session/session.routes';
+
+const routes = express.Router();
+
+routes.get('/ping', (req: Request, res: Response, next: NextFunction): any => {
```
Comment (danylkaaa) [inline src/modules/routes.ts]: move to liveops module

File: src/modules/routes.ts [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+  logger.info('Ping Check Request');
+  res.status(200).json({ message: 'Pong!' });
+});
+
+routes.use('/api/user', userRoutes)
```
Comment (danylkaaa) [inline src/modules/routes.ts]: /api/v1/user

File: src/modules/session/controller/refreshToken/refreshToken.controller.ts [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+    const decodedUser: IUser = TokenService.verifyToken(refreshToken, config.refreshToken);
+    const user: IUser | null = await UserService.getUserByEmail(decodedUser.email);
+
+    if (!user) {
+      throw new NotAuthorizedError('User not found');
+    }
+
+    const accessToken = TokenService.createToken(user);
+    const newRefreshToken = TokenService.refreshToken(user);
```
Comment (danylkaaa) [inline src/modules/session/controller/refreshToken/refreshToken.controller.ts]: inconsistent methods naming

File: src/modules/session/controller/refreshToken/refreshToken.controller.ts [hunk 19]
```diff
@@ -0,0 +1,31 @@
+import { type NextFunction, type Request, type Response } from 'express';
+import { NotAuthorizedError } from '../../../../errors';
+import { type IUser } from '../../../user/type';
+import TokenService from '../../../../services/TokenService';
+import config from '../../../../config/config';
+import UserService from '../../../user/user.service';
+
+const refreshToken = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+  try {
```
Comment (danylkaaa) [inline src/modules/session/controller/refreshToken/refreshToken.controller.ts]: you don't need try/catch at all. All errors will be handle by global error handler

File: src/modules/user/controllers/login/login.controller.ts [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+import TokenService from '../../../../services/TokenService';
+import { type IUser } from '../../type';
+import { NotAuthorizedError } from '../../../../errors';
+import { SessionService } from '../../../session';
+import { v4 as uuidv4 } from 'uuid';
+
+const login = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+  const { email, password } = req.body;
+
+  try {
```
Comment (danylkaaa) [inline src/modules/user/controllers/login/login.controller.ts:9]: try/catch is not needed

File: src/modules/user/controllers/login/login.controller.ts [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+    const user: IUser | null = await UserService.getUserByEmail(email);
```
Comment (danylkaaa) [inline src/modules/user/controllers/login/login.controller.ts]: implicit types no need here, TS will handle types automatically

File: src/modules/user/controllers/login/login.controller.ts [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+
+    if (!user) {
+      throw new NotAuthorizedError('User could not be found');
+    }
+
+    const isPasswordEqual = await EncryptionService.comparePassword(password, user.password);
+
+    if (!isPasswordEqual) {
+      throw new NotAuthorizedError('Wrong password');
+    }
```
Comment (danylkaaa) [inline src/modules/user/controllers/login/login.controller.ts]: this code should be implemented in UserService.getUserByPasswordAndEmail

Comment (danylkaaa) [reply src/modules/user/controllers/login/login.controller.ts]: and better to use passport.basicAuth/localAuth for finding user by email/psw and assigning it to request context

File: src/modules/user/controllers/login/login.controller.ts [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+
+    const sessionId = uuidv4();
+    const sessionExpiresDate = new Date(new Date().getTime() + (24 * 60 * 60 * 1000));
+
+    const session = await SessionService.createSession({
+      sid: sessionId,
+      userId: user.id,
+      expires: sessionExpiresDate,
+      data: JSON.stringify(token)
+    });
```
Comment (danylkaaa) [inline src/modules/user/controllers/login/login.controller.ts]: sessionId should be created in SessionService

File: src/modules/user/controllers/logout/logout.controller.ts [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+  const { sessionId } = req.body;
+
+  try {
+    if (!sessionId) {
+      throw new ValidationError('SessionId is undefined');
+    }
+
+    await SessionService.removeSession(sessionId);
+
+    res.status(200).json({ message: `Session ${sessionId as string} was removed` })
```
Comment (danylkaaa) [inline src/modules/user/controllers/logout/logout.controller.ts]: example of good controller

File: src/modules/user/controllers/signUp/signUp.controller.ts [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    const userResponse: Omit<IUser, 'password'> = {
+      firstName: user.firstName,
+      lastName: user.lastName,
+      role: user.role,
+      email: user.email
+    }
+
+    res.status(200).json({ user: userResponse });
```
Comment (danylkaaa) [inline src/modules/user/controllers/signUp/signUp.controller.ts]: userService should handle it

File: src/modules/user/user.service.ts [hunk 26]
```diff
@@ -0,0 +1,14 @@
+import { UserRepo } from './index';
+import { type IUser, type IUserInstance } from './type';
+
+class UserService {
```
Comment (danylkaaa) [inline src/modules/user/user.service.ts:6]: your implementation of UserService is too dummy. Instead, UserService should be responsible for all manipulation with user data. But you moved it to controller lvl, what is wrong

File: src/server.ts [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+  logger.info(`️[server]: Server is running at http://localhost:${config.port}`);
+});
+
+database.sync()
+  .then(() => {
+    console.log('Synced db.');
+  })
+  .catch((err: { message: string; }) => {
+    console.log('Failed to sync db: ' + err.message);
+  });
```
Comment (danylkaaa) [inline src/server.ts]: no, do not sync db

File: src/services/TokenService.ts [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+    return jwt.sign(
+      {
+        id: user.id,
+        firstName: user.firstName,
+        lastName: user.lastName,
+        role: user.role,
+        email: user.email
+      },
+      config.token,
+      { expiresIn: '3h' }
```
Comment (danylkaaa) [inline src/services/TokenService.ts]: time should be listed in server config

File: src/services/TokenService.ts [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+  static refreshToken (user: IUser): string {
+    return jwt.sign({
+      id: user.id,
+      firstName: user.firstName,
+      lastName: user.lastName,
+      role: user.role,
+      email: user.email
+    },
+    config.refreshToken,
+    { expiresIn: '5h' })
```
Comment (danylkaaa) [inline src/services/TokenService.ts]: same

File: migrations/20230412105758-create-sessions.js (hunk 6 above)
Comment (davidkin) [reply migrations/20230412105758-create-sessions.js:20]: Do you mean a DATETIME format? But Sequelize allow only DATE

File: src/modules/session/controller/refreshToken/refreshToken.controller.ts (hunk 19 above)
Comment (davidkin) [reply src/modules/session/controller/refreshToken/refreshToken.controller.ts]: how will these errors be passed to this error-middleware?

File: seeders/20230408144804-demo-user.js (hunk 9 above)
Comment (davidkin) [reply seeders/20230408144804-demo-user.js:6]: But this user already has a role:['admin']

File: migrations/20230412105758-create-sessions.js (hunk 6 above)
Comment (danylkaaa) [reply migrations/20230412105758-create-sessions.js:20]: oh, I got it. It's internal Sequalize impl. Sorry!

**Pull request merged**